### PR TITLE
feat: add bash completion

### PR DESCRIPTION
**Pull request by tomberek**
Note: this is just for the `nom <subcommand>` usage. But a similar trick can be used for `nom-build`.

Adds bash completion calling nix's and correcting the order of the args.

Another approach would be to call the normal bash nix completion but correct for the additional three arguments injected by nom, but this seems a bit cleaner to chain load the completion script.

Fixes: https://github.com/maralorn/nix-output-monitor/issues/31

### CODE REVIEW

File: exe/Main.hs [hunk 1]
```diff
@@ -48,7 +48,7 @@ replaceCommandWithExit :: [String] -> [String]
 replaceCommandWithExit = (<> ["--command", "sh", "-c", "exit"]) . takeWhile (\x -> x /= "--command" && x /= "-c")
 
 withJSON :: [String] -> [String]
-withJSON x = "-v" : "--log-format" : "internal-json" : x
+withJSON x = x <> ["-v", "--log-format", "internal-json"]
```
Comment (maralorn) [inline exe/Main.hs:51]: What is the purpose of this change?

From my understanding this means, that the user can’t override the debug-level anymore.

File: completions/completion.bash [hunk 2]
```diff
@@ -0,0 +1,2 @@
+__load_completion nix
+complete -F _complete_nix nom
```
Comment (maralorn) [inline completions/completion.bash:2]: Who would have thought that it’d be that simple.

**Comment by maralorn** [comment]
Thank you very much. This is a long-awaited improvement! I left a few comments.

File: exe/Main.hs (hunk 1 above)
Comment (tomberek) [reply exe/Main.hs:51]: The purpose is to make the NIX_GET_COMPLETIONS number correct. It is a number indicating which word of the input is being completed (eg: tabbing on a word in the middle of a command). Thus means we can use the Nix completion logic as-is.

Another approach could be to adjust that logic in the completion code with a +3 to account for the extra args.

Comment (maralorn) [reply exe/Main.hs:51]: The completion code has no clue how we will be calling Nix, right? I can’t imagine a completion that would behave differently because of those two extra arguments. I vote for not changing the Haskell, but leaving the completion like it is. That seems good enough.

The bigger problem is, that this will complete e.g. `nom copy` although nom does not support that.

Comment (tomberek) [reply exe/Main.hs:51]: The way the Nix completions works is that we pass in the argument number of the thing being completed.  So for example, you can run this directly to experiment...

```
[tom@tframe:~]$ NIX_GET_COMPLETIONS=2 nix build nixpkgs#hello
attrs
nixpkgs#hello
nixpkgs#hello-unfree
nixpkgs#hello-wayland
```
and it works if you have the cursor in the middle of the line; notice the "2".
```
[tom@tframe:~]$ NIX_GET_COMPLETIONS=2 nix build nixpkgs#hello --json -L
attrs
nixpkgs#hello
nixpkgs#hello-unfree
nixpkgs#hello-wayland
```

Because nom injects three extra arguments, the end result would be:
```
[tom@tframe:~]$ NIX_GET_COMPLETIONS=2 nix build -v --log-format internal-json nixpkgs#hello<TAB>
```

Which would complete the "-v" argument, not the expected one. The bash completer logic in _complete_nix does this. The alternative I mentioned (and was my original implementation) is to copy the nix completion logic, make a specific one for nom and modify it to adjust for the extra arguments:
https://github.com/NixOS/nix/blob/f0180487a0e4c0091b46cb1469c44144f5400240/misc/bash/completion.sh#L18

```diff
- done < <(NIX_GET_COMPLETIONS=$cword "${words[@]}" 2>/dev/null)
+ done < <(NIX_GET_COMPLETIONS=$(( cword + 3 )) "${words[@]}" 2>/dev/null)
```

Lastly, you can write your own completion handler, and call nix correctly yourself and write a completions/completion.bash to use that. This gives you the most control, but less re-use.

Comment (maralorn) [reply exe/Main.hs:51]: Yeah, but what I am saying is that the completion engine has no clue that we will be calling `nix build` with `-v --log-format internal-json`. When you type `nom build nixpkgs#hel<TAB>` it will use the completion for `nix build nixpkgs#hel` and then after the user selected nixpkgs#hello and pressed enter we can insert flags wherever we want. That can’t mess with the completion.

Comment (tomberek) [reply exe/Main.hs:51]: Ah! Good point.Then I don't know why I I had to add this. Perhaps left over from the previous style of implementation. I'll remove.

Comment (tomberek) [reply exe/Main.hs:51]: No, now i need to back on that statement. The original exec `nom` is called by _complete_nix bash function in `"${words[@]}`, so we do need a correction somewhere. I updated it to another approach where `nom` detects this case can calls the underlying nix without the extra args.

**Comment by maralorn** [comment]
Have you tested this successfully? When I test this I get the output below, where I think the noise with the "Finished at" will probably break completion?
At least completion didn’t work for me in a nix shell.
```
[maralorn@hephaistos:~]$ NIX_GET_COMPLETIONS=1 nix build pkgs#
normal
build	

[maralorn@hephaistos:~]$ NIX_GET_COMPLETIONS=1 nom build pkgs#
Finished at 22:26:31 after 0s
normal
build

**Comment by philiptaron** [comment]
I also was unable to get this to work at all, even with futzing to remove `runMonitoredCommand` when `NIX_GET_COMPLETIONS` was set. (The `Finished at 22:26:31 after 0s` is coming from the `runMonitoredCommand` code.)

**Comment by pdietl** [comment]
I'm going to attempt to take a stab at fixing this issue. Time to open up the Haskell book again.

**Comment by maralorn** [comment]
Done via #151, thx everyone!
